src/main/scala/me/rexim/morganey/interpreter/MorganeyExecutor.scala

```diff
@@ -0,0 +1,46 @@
+package me.rexim.morganey.interpreter
+
+import scala.util._
+
+import me.rexim.morganey.ast._
+import me.rexim.morganey.syntax.{LambdaParser, LambdaParserException}
+import me.rexim.morganey.module._
+import me.rexim.morganey.util._
+
+import java.io.Reader
+
+object MorganeyExecutor {
+
+  def interpretNode(node: MorganeyNode, moduleFinder: ModuleFinder): Try[List[MorganeyBinding]] =
+    node match {
+      case binding: MorganeyBinding => Success(List(binding))
+      case MorganeyLoading(Some(modulePath)) => loadModule(modulePath, moduleFinder)
+      case MorganeyLoading(None) => Success(List())
+    }
+
+  def loadModuleFromReader(reader: Reader, moduleFinder: ModuleFinder): Try[List[MorganeyBinding]] =
+    LambdaParser.parseAll(LambdaParser.module, reader) match {
+      case LambdaParser.Success(nodes, _) => {
+        sequence(nodes.map(interpretNode(_, moduleFinder))).map(_.flatten)
+      }
+      case LambdaParser.NoSuccess(message, _) => Failure(new LambdaParserException(message))
+    }
```

Comment:
Because there is an implicit for handling parse-results more easily, this could be rewritten as

```scala
def loadModuleFromReader(reader: Reader, moduleFinder: ModuleFinder): Try[List[MorganeyBinding]] =
  LambdaParser.parseWith(reader, _.module) map {
    nodes => sequence(nodes.map(interpretNode(_, moduleFinder))).map(_.flatten)
  }
```